### PR TITLE
feat(background): BGM-PR-03 config + AppContext + memory repository

### PR DESCRIPTION
**Pull request by slin1237**
## Summary

Third PR in the background-mode series. Scaffolds the gateway for background responses without enabling the create/cancel/stream handlers (those land in BGM-PR-04+). Ships the first concrete \`BackgroundResponseRepository\` implementation (memory) so dev/CI/single-node deployments can exercise the trait end-to-end.

Closes the acceptance criteria for BGM-PR-03 in \`.claude/designs/background-mode/2026-04-17-background-mode-task-breakdown.md\`.

## What changed

**crates/data_connector/src/memory_background.rs** (new, ~880 LOC)
- Full \`MemoryBackgroundRepository\` — all 10 \`BackgroundResponseRepository\` methods implemented with a single \`Arc<RwLock<State>>\`.
- State machine: Queued → InProgress (claim) → Completed / Incomplete / Failed / Cancelled (finalize).
- Cancel/complete race resolution via \`cancel_won\` in \`FinalizeResult\`.
- Monotonic per-response stream sequence allocated atomically in \`append_stream_event\`.
- Lease fencing in \`heartbeat\` and \`finalize\`; expired leases requeued with bumped \`retry_attempt\`.
- Explicit non-durability: process restart returns \`NotFound\` for prior IDs.
- 14 unit tests covering every state-machine path.

**crates/data_connector/src/factory.rs**
- \`StorageBundle.background_repository: Option<Arc<dyn BackgroundResponseRepository>>\`.
- Memory: \`Some(MemoryBackgroundRepository::new())\`.
- Postgres / Oracle / Redis / None: \`None\` (durable impls land in BGM-PR-05 / PR-06; Redis out of scope for Phase 1).

**model_gateway/src/config/types.rs**
- \`BackgroundConfig\` struct with 9 operational tuning knobs.
- **No \`background_enabled\` flag** — backend capability (repository \`Some\` / \`None\`) is the gate. Simpler surface than the original design.
- Durations as \`u64\` seconds / milliseconds + accessor methods for \`Duration\` (no new humantime dep).
- Defaults match the design doc (16 workers, 10k queue, 60s lease, 3 retries, 2s→60s backoff, 30s sweep, 500ms poll, 15m stream retention).
- Backward-compatible serde: existing YAML configs keep deserializing unchanged.

**model_gateway/src/app_context.rs**
- \`AppContext.background_repository\` carried via \`AppContextBuilder\` from the factory bundle.

**model_gateway/src/routers/common/background/mod.rs** (new)
- \`BackgroundServices\` placeholder that BGM-PR-04 will populate with shared create / cancel / resume helpers.

**model_gateway/src/routers/grpc/regular/responses/handlers.rs**
- Existing hard rejection of \`background=true\` is preserved with a pointer to BGM-PR-04 where it gets replaced with delegation.
- **No client-visible behavior change in this PR.**

## Why

Avoid shaping the surface area in-flight. Get the config, AppContext wiring, factory, and module boundaries right once; then PR-04 just fills in the create handler, PR-05 the Postgres backend, PR-06 the Oracle backend. Ships memory alongside to give the trait shape an early stress test and unblock dev/CI workflows.

Memory backend was added per user direction after initial scoping: dev, CI, and single-node deployments benefit from background responses without needing a Postgres container. Non-durability is explicit in the module doc — consistent with how \`MemoryResponseStorage\` already treats synchronous responses.

## How

\`background_enabled\` flag dropped during review after the observation that backend capability already gates availability — having both is redundant and creates a misconfiguration surface where the flag contradicts the repository presence.

Fixed-name constants (\`BG_QUEUE_*\`, \`STREAM_CHUNKS_*\`) on the durable migrations in BGM-PR-02 already made the repository boundary explicit; this PR just carries that pattern into the gateway layer via \`BackgroundServices\`.

## Test plan

- [x] \`cargo +nightly fmt --all\` — clean
- [x] \`cargo clippy --all-targets --all-features -- -D warnings\` — clean
- [x] \`cargo test -p data-connector --lib\` — **269 passed**, 0 failed (14 new for \`MemoryBackgroundRepository\`)
- [x] \`cargo test -p smg --lib\` — **601 passed**, 0 failed (3 new for \`BackgroundConfig\`)
- [x] Verified Python bindings use \`RouterConfig::builder()\` — no struct literal to update

Refs: BGM-PR-01 #1244, BGM-PR-02 #1258

## Summary by CodeRabbit

* **New Features**
  * Background task processing system for asynchronous job execution with automatic queueing, worker lease management, heartbeat monitoring, configurable retry logic, and cancellation support.

### CODE REVIEW

File: crates/data_connector/src/memory_background.rs [hunk 1]
```diff
@@ -0,0 +1,884 @@
...
+                cancel_won: false,
+            });
+        }
+
+        // Lease fencing: only the worker that holds the lease can finalize.
+        if entry.worker_id.as_deref() != Some(update.worker_id.as_str()) {
+            return Err(BackgroundRepositoryError::LeaseNotHeld {
+                response_id: update.response_id.clone(),
+                worker_id: update.worker_id,
+            });
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/data_connector/src/memory_background.rs]: Fixed in `335a112e`. `finalize` now takes `now: DateTime<Utc>` and rejects with `LeaseNotHeld` when `lease_expires_at <= now` — even if `requeue_expired` hasn't swept yet. Stale workers can't commit terminal state after their lease window. New regression test `finalize_rejects_expired_lease_even_before_sweeper`.

File: crates/data_connector/src/memory_background.rs [hunk 2]
```diff
@@ -0,0 +1,884 @@
...
+
+        // Cancel-before-claim: the design says a claim on a cancel-requested
+        // row transitions straight to Cancelled without invoking the worker.
+        // Expose this as `claim_next returns None` from the worker's POV so
+        // the worker never sees a job it must not execute; the cancel path
+        // is what actually observed the transition.
+        if entry.cancel_requested {
+            entry.status = InternalStatus::Cancelled;
+            entry.completed_at = Some(now);
+            return Ok(None);
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/data_connector/src/memory_background.rs]: Fixed in `335a112e`. `claim_next` now loops after auto-cancelling a queued row with `cancel_requested=true` instead of returning `None`. Regression test `claim_next_skips_cancelled_and_returns_next_runnable` walks the full edge case (InProgress + cancel → requeue_expired → Queued+cancel_requested → next claim skips it to return the next runnable row).

File: model_gateway/src/routers/common/background/mod.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+use crate::config::BackgroundConfig;
+
+/// Handle carrying the background-mode repository and operational tuning
+/// into request handlers. Cheap to clone (internally `Arc`).
+///
+/// Stored on [`crate::app_context::AppContext`] as
+/// `Option<BackgroundServices>`; `None` means the active history backend
+/// does not support background mode and request handlers must reject
+/// `background=true` with `400 background_not_supported`.
```
[the remark this answers is not in the document]
Comment (slin1237) [reply model_gateway/src/routers/common/background/mod.rs]: Already fixed — the stale "Stored on AppContext as Option<BackgroundServices>" comment was dropped during the comment-cleanup pass. Current doc on `BackgroundServices` just says "BGM-PR-03 ships only BackgroundServices; concrete handlers land in later PRs" and no longer makes any claim about AppContext storage.

File: crates/data_connector/src/memory_background.rs [hunk 4]
```diff
@@ -0,0 +1,884 @@
...
+        }
+
+        let expires_at =
+            now + chrono::Duration::from_std(lease).unwrap_or(chrono::Duration::zero());
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/data_connector/src/memory_background.rs]: Fixed in `335a112e`. Replaced `unwrap_or(chrono::Duration::zero())` in both sites with a `to_chrono` helper that returns `Err(BackgroundRepositoryError::Backend(...))` on overflow. No silent fallback to an instant-expiring lease on misconfig.

File: crates/data_connector/src/memory_background.rs [hunk 5]
```diff
@@ -0,0 +1,884 @@
...
+        for _ in 0..3 {
+            let id = repo
+                .claim_next("w1", now, lease)
+                .await
+                .unwrap()
+                .map(|j| j.response_id.0)
+                .unwrap_or_default();
+            claim_order.push(id);
+        }
+        assert_eq!(claim_order, vec!["r2", "r1", "r3"]);
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/data_connector/src/memory_background.rs]: Fixed in `335a112e`. Test renamed to `claim_next_respects_priority_order` and given fully distinct priorities (1 / 5 / 10). Ordering is now determined by the priority tiebreaker alone — no dependency on `Utc::now()` returning distinct nanoseconds under coarse clocks.

File: model_gateway/src/config/types.rs [hunk 6]
```diff
@@ -47,6 +132,10 @@ pub struct RouterConfig {
     pub storage_context_headers: HashMap<String, String>,
     #[serde(default)]
     pub memory_runtime: MemoryRuntimeConfig,
+    /// Operational tuning for background mode. Availability is gated by
+    /// backend capability (memory / postgres / oracle), not by a flag here.
+    #[serde(default)]
+    pub background: BackgroundConfig,
```
[the remark this answers is not in the document]
Comment (slin1237) [reply model_gateway/src/config/types.rs:110]: Fixed in `335a112e`. `ConfigValidator::validate_background` rejects zero values on `worker_concurrency`, `max_queue_depth`, `lease_duration_secs`, `sweep_interval_secs`, `poll_interval_ms`, `stream_retention_secs`, and enforces `retry_base_delay_secs <= retry_max_delay_secs`. Two regression tests pin the behavior.

File: crates/data_connector/src/memory_background.rs [hunk 7]
```diff
@@ -0,0 +1,809 @@
...
+            return Ok(StoredCancelResult::AlreadyTerminal);
+        }
+
+        match entry.status {
+            InternalStatus::Queued => {
+                entry.status = InternalStatus::Cancelled;
+                entry.completed_at = Some(now);
+                entry.worker_id = None;
+                entry.lease_expires_at = None;
+                Ok(StoredCancelResult::QueuedCancelled)
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/data_connector/src/memory_background.rs:287]: Fixed in follow-up PR #1348. `request_cancel` now overwrites `raw_response.status` to `"cancelled"` via a shared `overwrite_raw_response_status` helper after flipping the queued row. Regression test `queued_cancel_rewrites_raw_response_status`.

File: crates/data_connector/src/memory_background.rs [hunk 8]
```diff
@@ -0,0 +1,809 @@
...
+        }
+
+        let (final_status, cancel_won) = if entry.cancel_requested {
+            (FinalizeStatus::Cancelled, true)
+        } else {
+            (update.status, false)
+        };
+
+        entry.status = InternalStatus::from_finalize(final_status);
+        entry.raw_response = update.raw_response;
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/data_connector/src/memory_background.rs:391]: Fixed in follow-up PR #1348. `finalize` with `cancel_won=true` now calls the same `overwrite_raw_response_status` helper after assigning `update.raw_response`, so worker-supplied `status: "completed"` / `"failed"` payloads get overwritten to `"cancelled"` when cancel wins. `finalize_cancel_wins_over_worker_status` was extended to assert the patched payload.
